tensornetwork/linalg/krylov.py

```diff
@@ -0,0 +1,326 @@
+  The solution can in this case be repeatedly
+  improved, to a point, by restarting the Arnoldi iterations each time
+  `num_krylov_vectors` is reached. Unfortunately the optimal parameter choices
+  balancing expense and accuracy are difficult to predict in advance, so
+  applying this function requires a degree of experimentation.
+
+  In a tensor network code one is typically interested in A_mv implementing
+  some tensor contraction. This implementation thus allows `b` and `x0` to be
+  of whatever arbitrary, though identical, shape `b = A_mv(x0, ...)` expects.
+  Reshaping to and from a matrix problem is handled internally.
+
+  Args:
+    A_mv     : A function `v0 = A_mv(v, *A_args, **A_kwargs)` where `v0` and
+               `v` have the same shape.
+    b        : The `b` in `A @ x = b`; it should be of the shape `A_mv`
+               operates on.
+    A_args   : Positional arguments to `A_mv`, supplied to this interface
+               as a list.
+               Default: None.
+    x0       : An optional guess solution. Zeros are used by default.
+               If `x0` is supplied, its shape and dtype must match those of
+               `b`, or an
+               error will be thrown.
+               Default: zeros.
+    tol, atol: Solution tolerance to achieve,
+               norm(residual) <= max(tol*norm(b), atol).
+               Default: tol=1E-05
+                        atol=tol
+    num_krylov_vectors
+             : Size of the Krylov space to build at each restart.
+               Expense is cubic in this parameter. If supplied, it must be
+               an integer in 0 < num_krylov_vectors <= b.size.
+               Default: b.size.
+    maxiter  : The Krylov space will be repeatedly rebuilt up to this many
+               times. Large values of this argument
+               should be used only with caution, since especially for nearly
+               symmetric matrices and small `num_krylov_vectors` convergence
+               might well freeze at a value significantly larger than `tol`.
+               Default: 1.
+    M        : Inverse of the preconditioner of A; see the docstring for
+               `scipy.sparse.linalg.gmres`. This is only supported in the
+               numpy backend. Supplying this argument to other backends will
+               trigger NotImplementedError.
+               Default: None.
+
+  Raises:
+    ValueError: -if `x0` is supplied but its shape differs from that of `b`.
+                -in NumPy, if the ARPACK solver reports a breakdown (which
+                 usually indicates some kind of floating point issue).
+                -if num_krylov_vectors is 0 or exceeds b.size.
+                -if tol was negative.
+                -if M was supplied with any backend but NumPy.
+
+  Returns:
+    x       : The converged solution. It has the same shape as `b`.
+    info    : 0 if convergence was achieved, the number of restarts otherwise.
+  """
+  try:
```

Comment:
why test this here and not in the other routines?

Reply:
In the other cases the analogous tests are done by krylov_error_checks (but GMRES has additional data those tests don't cover).